# Agents and Tools

Agents involve an LLM making decisions about which Actions to take, taking that Action, seeing an Observation, and repeating that until done.


When used correctly agents can be extremely powerful. In order to load agents, you should understand the following concepts:

- Tool: A function that performs a specific duty. This can be things like: Google Search, Database lookup, Python REPL, other chains.
- LLM: The language model powering the agent.
- Agent: The agent to use.


For example I have to travel from Dubai to Canada, I type this in ChatGPT

Give me two flight options from Dubai to Canada on September 1, 2024 | ChatGPT will not be able to answer because has knowledge till
September 2021

ChatGPT plus has Expedia Plugin, if we enable this plugin it will go to Expedia Plugin and will try to pull information about Flights & it will show the information

When we think about LLM. Many people think that it is just a knowledge engine, it has knowledge and it will try to give answer based on that knowledge but the knowledge is only limited to September 2021. The think that most people missout is that Large Lanaguage Model has a reasoning engine, and using that reasoning engine it can figure out when someone types this type of Question


Give me  two flight options from Dubai to Canada on September 1, 2024


As a human we go to Expedia as we have a reasoning engine in our brain.

LLM has a reasoning engine as well, so it will figure out the Source, Destination, Date and it will call Expedia Plugin and it will return response back.

#**How much is US GDP in 2024? plus 5**


# Agent will conenct with external tools and it will use LLM reasoning capabilities

# To access Google Search Results in Real Time we use serpapi

# serpapi and llm-math tool

In [ ]:
!pip install google-search-results

  Preparing metadata (setup.py) ... done
  Created wheel for google-search-results: filename=google_search_results-2.4.2-py3-none-any.whl size=32004 sha256=01c8d4136f0787aeba3d27fb3794d52b87ae8ab9cd6211ea9ec550b4f43ae9b2
  Stored in directory: /root/.cache/pip/wheels/d3/b2/c3/03302d12bb44a2cdff3c9371f31b72c0c4e84b8d2285eeac53
Successfully built google-search-results


In [ ]:
!pip install langchain

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 974.6/974.6 kB 12.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 321.8/321.8 kB 24.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.1/127.1 kB 14.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 145.0/145.0 kB 13.3 MB/s eta 0:00:00


In [ ]:
!pip install -U langchain-community

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 19.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 5.4 MB/s eta 0:00:00


In [ ]:
%pip install -qU langchain-openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 kB 1.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 327.4/327.4 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 41.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 8.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 5.4 MB/s eta 0:00:00


In [ ]:
import os

In [ ]:
os.environ["SERPAPI_API_KEY"]='fc00bae5ada1188e2edcb51aadea4b6a6ff7dad69dc5fcd4260ddba892547851'

In [ ]:
from google.colab import userdata
OPENAI_API_KEY=userdata.get('OPENAI_API_KEY')

In [ ]:
os.environ["OPENAI_API_KEY"]=OPENAI_API_KEY

In [ ]:
from langchain.agents import AgentType,initialize_agent, load_tools

In [ ]:
from langchain_openai import ChatOpenAI

In [ ]:
llm = ChatOpenAI(
    model="gpt-3.5-turbo",
    temperature=0)

In [ ]:
tool = load_tools(["serpapi","llm-math"],llm=llm)

In [ ]:
agent1 = initialize_agent(tool, llm, agent= AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose= True)

In [ ]:
agent1.run("How much is US GDP in 2024 plus 5?")

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(




> Entering new AgentExecutor chain...
I need to calculate the US GDP in 2024 and then add 5 to it.
Action: Calculator
Action Input: US GDP in 2024
Observation: Answer: 21433000000000.0
Thought:Now I need to add 5 to the US GDP in 2024.
Action: Calculator
Action Input: 21433000000000.0 + 5
Observation: Answer: 21433000000005.0
Thought:I now know the final answer.
Final Answer: US GDP in 2024 plus 5 is 21,433,000,000,005.

> Finished chain.


'US GDP in 2024 plus 5 is 21,433,000,000,005.'

In [ ]:
agent1.run("in which year indian team won its second world cup give me all the player name of the winning team??")



> Entering new AgentExecutor chain...
I need to search for the year when the Indian team won its second World Cup and then find the names of all the players in that winning team.
Action: Search
Action Input: "Indian cricket team second World Cup win year"
Observation: India's first World Cup victory came in 1983 when they defeated the West Indies in the final held at Lord's Cricket Ground in London, England. Their second World Cup win was in 2011, when they won the tournament on home soil, defeating Sri Lanka in the final at Wankhede Stadium in Mumbai, India.
Thought:Now that I know the Indian team won its second World Cup in 2011, I need to search for the names of all the players in the winning team.
Action: Search
Action Input: "Indian cricket team 2011 World Cup winning squad"
Observation: ['Checkout team and player detaills of India Squad - ICC Cricket World Cup, 2011 Squad details on ESPNcricinfo.com.', 'Indian innings ; Sri Lanka. 274/6 (50 overs). v. India 277/4 (48.2 overs) ;

"The players in the Indian cricket team's 2011 World Cup winning squad were MS Dhoni (c), Virender Sehwag, Sachin Tendulkar, Gautam Gambhir, Virat Kohli, Yuvraj Singh, Suresh Raina, and Harbhajan Singh."

In [ ]:
agent.run("How much is india GDP in 2024 plus 5?")



> Entering new AgentExecutor chain...
I need to find the GDP of India in 2024 and then add 5 to it.
Action: Search
Action Input: "India GDP 2024"
Observation: {'type': 'organic_result', 'title': 'Economy of India', 'source': 'Wikipedia'}
Thought:I need to click on the Wikipedia link to find the specific GDP information for 2024.
Action: Search
Action Input: "India GDP 2024 Wikipedia"
Observation: $4.038 trillion (nominal; 2024 est.) $14.594 trillion (PPP; 2024 est.)
Thought:Now I need to add 5 to the GDP of India in 2024.
Action: Calculator
Action Input: 4.038 trillion + 5
Observation: Answer: 4038000000005.0005
Thought:I now know the final answer
Final Answer: $4,038,000,000,005.00

> Finished chain.


'$4,038,000,000,005.00'

In [ ]:
agent.run("who won the india's 2024 general election?")



> Entering new AgentExecutor chain...
I should use the search tool to find the answer to this question.
Action: Search
Action Input: "India 2024 general election winner"
Observation: ["Prime Minister Narendra Modi's National Democratic Alliance coalition has won a majority in parliament, according to official results from ...", 'General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', 'The governing Bharatiya Janata Party (BJP) of Prime Minister Narendra Modi won 240 seats, falling short of the 272-mark that signifies a ...', "See results and maps from India's 2024 Lok Sabha elections.", 'Indian PM Narendra Modi has won a third consecutive term in a much tighter general election than anticipated. His Bharatiya Janata Party ...', 'The successful execution of these phases was a testament to the dedication of voters, political parties, candidates, election personnel, and ...', "Vote cou

"Prime Minister Narendra Modi's National Democratic Alliance coalition won the India's 2024 general election."

In [ ]:
agent.run("how much seat BJP won in india's general election in 2024 plus 10000?")



> Entering new AgentExecutor chain...
I need to find out how many seats BJP won in India's general election in 2024 first before adding 10000.
Action: Search
Action Input: "BJP seats won in India general election 2024"
Observation: ['General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', "Prime Minister Narendra Modi's National Democratic Alliance coalition has won a majority in parliament, according to official results from ...", 'The Bharatiya Janata Party, together with its National Democratic Alliance, has won a majority.', 'Prime Minister Narendra Modi and his Bharatiya Janata Party (BJP) won the most seats for the third consecutive time, but have fallen short of a ...', "See results and maps from India's 2024 Lok Sabha elections.", 'The governing Bharatiya Janata Party (BJP) of Prime Minister Narendra Modi won 240 seats, falling short of the 272-mark that signifies a ...', 'E

'10240'

In [ ]:
!pip install wikipedia

  Preparing metadata (setup.py) ... done
  Created wheel for wikipedia: filename=wikipedia-1.4.0-py3-none-any.whl size=11680 sha256=0511c5010f1f455065eef86a1a572880ff08eaddbc8102c327a74abdee351494
  Stored in directory: /root/.cache/pip/wheels/5e/b6/c5/93f3dec388ae76edc830cb42901bb0232504dfc0df02fc50de
Successfully built wikipedia


In [ ]:
tool = load_tools(["wikipedia","llm-math"],llm=llm)

In [ ]:
agent = initialize_agent(
    tool,
    llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    verbose=True,

)

In [ ]:
agent.run("in which year first indian film released?")



> Entering new AgentExecutor chain...
I need to find out when the first Indian film was released.
Action: wikipedia
Action Input: First Indian film
Observation: Page: Cinema of India
Summary: The Cinema of India, consisting of motion pictures made by the Indian film industry, has had a large effect on world cinema since the second half of the 20th century. Indian cinema is made up of various film industries, including Hindi cinema, which makes motion pictures in the Hindi language and is one of the biggest film industries in the country. In 2021, Telugu cinema became the largest film industry in India in terms of box office receipts. Major centers of film production across the country include Mumbai, Chennai, Hyderabad,  Kolkata, Kochi, Bangalore, Bhubaneswar-Cuttack, and Guwahati. For a number of years, the Indian film industry has ranked first in the world in terms of annual film output. In 2022, Indian cinema earned ₹15,000 crore ($1.7955 billion) at the box-office.
Indian cinema 

'The first Indian film was released in 1913.'

In [ ]:
agent.run("give me a detail analysis of indian election 2024")



> Entering new AgentExecutor chain...
I should use Wikipedia to find information about the Indian election in 2024.
Action: wikipedia
Action Input: Indian election 2024
Observation: Page: 2024 Indian general election
Summary: General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the result was declared on 4 June to form the 18th Lok Sabha. On 7 June 2024, Modi confirmed the support of 293 MPs to Droupadi Murmu, the president of India. This marked Modi's third term as prime minister and his first time heading a coalition government, with the Telugu Desam Party of Andhra Pradesh and Janata Dal (United) of Bihar emerging as two main allies.
More than 968 million people out of a population of 1.4 billion people were eligible to vote, equivalent to 70 percent of the total population. 642 million voters participated in the election and 312 million of them were women, making it the highest ever pa

'The 2024 Indian general election was held in seven phases from 19 April to 1 June 2024 to elect all 543 members of the Lok Sabha. Narendra Modi secured his third term as prime minister, leading a coalition government with the support of 293 MPs. The election saw a record number of eligible voters and high participation, with the BJP losing its majority in the Lok Sabha.'

The 2024 Indian general election was held in seven phases from 19 April to 1 June 2024 to elect all 543 members of the Lok Sabha. Narendra Modi secured his third term as prime minister, leading a coalition government with the support of 293 MPs. The election saw a record number of eligible voters and high participation, with the BJP losing its majority in the Lok Sabha.